runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/engine/internal/HttpClientMetrics.kt

```diff
@@ -39,6 +44,7 @@ public object HttpClientMetricAttributes {
 public class HttpClientMetrics(
     scope: String,
     public val provider: TelemetryProvider,
+    threadStateCallback: (() -> ThreadState?)? = null,
```

Comment:
This seems different for no apparent reason. It's not wrong I guess but feels odd that we would special case it since each engine can adapt however they want to the current way metrics are exposed.

Reply:
Yes I concede it's very different from what we have elsewhere and feels very specialized right now.

I think one problem is that we don't directly control the thread pool. We don't decide when new threads are created or when they swap from idle to active. All we can do is inspect the state at any given moment to know the current value. That sounds like what an `AsyncUpDownCounter` is intended to do via a callback.

We could start a background thread in the OkHttp engine to periodically inspect the thread pool and stuff the idle/active counts into `HttpClientMetrics`, which would then be read when the callback is run. But this means the value will be out of sync by whatever interval the background thread uses. If, for instance, the background thread runs every 10 seconds but the meter polls every 5 seconds, then up to two datapoints will be reported inaccurately.

Is your greater concern that this change exposes the concept of callbacks outside of `HttpClientMetrics` and inverts the collection model from a push to a pull? Is it the inconsistency of some metrics being exposed outside of `HttpClientMetrics` as callbacks while others are mutable properties inside? Or is it that this may not be flexible enough to adapt to other engines where thread updates are more deterministic?